File: clients/go/node/p2p/compactblock.go
package p2p

import (
	"encoding/binary"
	"fmt"

	"rubin.dev/node/consensus"
	"rubin.dev/node/crypto"
)

const (
	CompactBlockShortIDBytes = 6
	compactBlockKeyDomain    = "RUBIN-CMPCT-v1"
)

type SendCmpctPayload struct {
	Announce        uint8
	ShortIDWTXID    uint8 // MUST be 1 in RUBIN
	ProtocolVersion uint32
}

func EncodeSendCmpctPayload(p SendCmpctPayload) ([]byte, error) {
	if p.ShortIDWTXID != 1 {
		return nil, fmt.Errorf("p2p: sendcmpct: shortid_wtxid must be 1")
	}
	out := make([]byte, 0, 6)
	out = append(out, p.Announce)
	out = append(out, p.ShortIDWTXID)
	var tmp4 [4]byte
	binary.LittleEndian.PutUint32(tmp4[:], p.ProtocolVersion)
	out = append(out, tmp4[:]...)
	return out, nil
}

func DecodeSendCmpctPayload(b []byte) (*SendCmpctPayload, error) {
	if len(b) != 6 {
		return nil, fmt.Errorf("p2p: sendcmpct: length mismatch")
	}
	p := &SendCmpctPayload{
		Announce:        b[0],
		ShortIDWTXID:    b[1],
		ProtocolVersion: binary.LittleEndian.Uint32(b[2:6]),
	}
	if p.ShortIDWTXID != 1 {
		return nil, fmt.Errorf("p2p: sendcmpct: shortid_wtxid must be 1")
	}
	return p, nil
}

type PrefilledTx struct {
	Index   uint64
	TxBytes []byte // canonical TxBytes (includes witness; includes DA payload in wire v2)
}

type CmpctBlockPayload struct {
	Header    consensus.BlockHeader
	Nonce     uint64
	TxCount   uint64
	ShortIDs  [][CompactBlockShortIDBytes]byte
	Prefilled []PrefilledTx
}

func EncodeCmpctBlockPayload(p CmpctBlockPayload) ([]byte, error) {
	if p.TxCount == 0 {
		return nil, fmt.Errorf("p2p: cmpctblock: tx_count must be >= 1")
	}
	if uint64(len(p.ShortIDs))+uint64(len(p.Prefilled)) != p.TxCount {
		return nil, fmt.Errorf("p2p: cmpctblock: shortid_count+prefilled_count must equal tx_count")
	}

	out := make([]byte, 0, BlockHeaderBytesLen+8+64)
	out = append(out, consensus.BlockHeaderBytes(p.Header)...)
	var tmp8 [8]byte
	binary.LittleEndian.PutUint64(tmp8[:], p.Nonce)
	out = append(out, tmp8[:]...)
	out = append(out, encodeCompactSize(p.TxCount)...)
	out = append(out, encodeCompactSize(uint64(len(p.ShortIDs)))...)
	for _, sid := range p.ShortIDs {
		out = append(out, sid[:]...)
	}

	out = append(out, encodeCompactSize(uint64(len(p.Prefilled)))...)
	var prevIdx uint64
	for i, pf := range p.Prefilled {
		if pf.Index >= p.TxCount {
			return nil, fmt.Errorf("p2p: cmpctblock: prefilled index out of range")
		}
		if i == 0 {
			out = append(out, encodeCompactSize(pf.Index)...)
		} else {
			if pf.Index <= prevIdx {
				return nil, fmt.Errorf("p2p: cmpctblock: prefilled indices not strictly increasing")
			}
			out = append(out, encodeCompactSize(pf.Index-prevIdx-1)...)
		}
		prevIdx = pf.Index
		// Ensure the bytes are parseable to avoid ambiguous concatenations.
		if _, err := consensus.ParseTxBytes(pf.TxBytes); err != nil {
			return nil, fmt.Errorf("p2p: cmpctblock: invalid prefilled tx_bytes: %w", err)
		}
		out = append(out, pf.TxBytes...)
	}
	return out, nil
}

func DecodeCmpctBlockPayload(b []byte) (*CmpctBlockPayload, error) {
	if len(b) < BlockHeaderBytesLen+8+1 {
		return nil, fmt.Errorf("p2p: cmpctblock: short payload")
	}
	off := 0
	headerBytes := b[off : off+BlockHeaderBytesLen]
	off += BlockHeaderBytesLen
	h, err := consensus.ParseBlockHeaderBytes(headerBytes)
	if err != nil {
		return nil, err
	}
	nonce := binary.LittleEndian.Uint64(b[off : off+8])
	off += 8

	txCount, used, err := readCompactSize(b[off:])
	if err != nil {
		return nil, err
	}
	off += used
	if txCount < 1 {
		return nil, fmt.Errorf("p2p: cmpctblock: tx_count must be >= 1")
	}

	shortidCount, used, err := readCompactSize(b[off:])
	if err != nil {
		return nil, err
	}
	off += used
	if shortidCount > txCount {
		return nil, fmt.Errorf("p2p: cmpctblock: shortid_count exceeds tx_count")
	}
	needSIDs := int(shortidCount) * CompactBlockShortIDBytes
	if len(b) < off+needSIDs {
		return nil, fmt.Errorf("p2p: cmpctblock: shortids truncated")
	}
	shortids := make([][CompactBlockShortIDBytes]byte, 0, int(shortidCount))
	for i := 0; i < int(shortidCount); i++ {
		var sid [CompactBlockShortIDBytes]byte
		copy(sid[:], b[off:off+CompactBlockShortIDBytes])
		shortids = append(shortids, sid)
		off += CompactBlockShortIDBytes
	}

	prefilledCount, used, err := readCompactSize(b[off:])
	if err != nil {
		return nil, err
	}
	off += used
	if prefilledCount > txCount {
		return nil, fmt.Errorf("p2p: cmpctblock: prefilled_count exceeds tx_count")
	}
	if shortidCount+prefilledCount != txCount {
		return nil, fmt.Errorf("p2p: cmpctblock: shortid_count+prefilled_count must equal tx_count")
	}

	prefilled := make([]PrefilledTx, 0, int(prefilledCount))
	var prevIdx uint64
	for i := 0; i < int(prefilledCount); i++ {
		delta, u, err := readCompactSize(b[off:])
		if err != nil {
			return nil, err
		}
		off += u
		var idx uint64
		if i == 0 {
			idx = delta
		} else {
			idx = prevIdx + 1 + delta
		}
		if idx >= txCount {
			return nil, fmt.Errorf("p2p: cmpctblock: prefilled index out of range")
		}
		if i > 0 && idx <= prevIdx {
			return nil, fmt.Errorf("p2p: cmpctblock: prefilled indices not strictly increasing")
		}
		prevIdx = idx

		tx, usedBytes, err := consensus.ParseTxBytesPrefix(b[off:])
		if err != nil {
			return nil, err
		}
		_ = tx // parse validates canonical structure
		txBytes := append([]byte(nil), b[off:off+usedBytes]...)
		off += usedBytes
		prefilled = append(prefilled, PrefilledTx{Index: idx, TxBytes: txBytes})
	}

	if off != len(b) {
		return nil, fmt.Errorf("p2p: cmpctblock: trailing bytes")
	}

	return &CmpctBlockPayload{
		Header:    h,
		Nonce:     nonce,
		TxCount:   txCount,
		ShortIDs:  shortids,
		Prefilled: prefilled,
	}, nil
}

type GetBlockTxnPayload struct {
	BlockHash [32]byte
	Indices   []uint64
}

func EncodeGetBlockTxnPayload(p GetBlockTxnPayload) ([]byte, error) {
	if len(p.Indices) == 0 {
		return nil, fmt.Errorf("p2p: getblocktxn: empty indices")
	}
	out := make([]byte, 0, 32+9+len(p.Indices)*5)
	out = append(out, p.BlockHash[:]...)
	out = append(out, encodeCompactSize(uint64(len(p.Indices)))...)
	var prev uint64
	for i, idx := range p.Indices {
		if i == 0 {
			out = append(out, encodeCompactSize(idx)...)
		} else {
			if idx <= prev {
				return nil, fmt.Errorf("p2p: getblocktxn: indices not strictly increasing")
			}
			out = append(out, encodeCompactSize(idx-prev-1)...)
		}
		prev = idx
	}
	return out, nil
}

func DecodeGetBlockTxnPayload(b []byte) (*GetBlockTxnPayload, error) {
	if len(b) < 32+1 {
		return nil, fmt.Errorf("p2p: getblocktxn: short payload")
	}
	var h [32]byte
	copy(h[:], b[:32])
	off := 32
	n, used, err := readCompactSize(b[off:])
	if err != nil {
		return nil, err
	}
	off += used
	if n == 0 {
		return nil, fmt.Errorf("p2p: getblocktxn: empty indices")
	}
	indices := make([]uint64, 0, int(n))
	var prev uint64
	for i := 0; i < int(n); i++ {
		delta, u, err := readCompactSize(b[off:])
		if err != nil {
			return nil, err
		}
		off += u
		var idx uint64
		if i == 0 {
			idx = delta
		} else {
			idx = prev + 1 + delta
		}
		if i > 0 && idx <= prev {
			return nil, fmt.Errorf("p2p: getblocktxn: indices not strictly increasing")
		}
		prev = idx
		indices = append(indices, idx)
	}
	if off != len(b) {
		return nil, fmt.Errorf("p2p: getblocktxn: trailing bytes")
	}
	return &GetBlockTxnPayload{BlockHash: h, Indices: indices}, nil
}

type BlockTxnPayload struct {
	BlockHash [32]byte
	Txs       [][]byte // each is canonical TxBytes
}

func EncodeBlockTxnPayload(p BlockTxnPayload) ([]byte, error) {
	out := make([]byte, 0, 32+9)
	out = append(out, p.BlockHash[:]...)
	out = append(out, encodeCompactSize(uint64(len(p.Txs)))...)
	for _, txb := range p.Txs {
		if _, err := consensus.ParseTxBytes(txb); err != nil {
			return nil, fmt.Errorf("p2p: blocktxn: invalid tx_bytes: %w", err)
		}
		out = append(out, txb...)
	}
	return out, nil
}

func DecodeBlockTxnPayload(b []byte) (*BlockTxnPayload, error) {
	if len(b) < 32+1 {
		return nil, fmt.Errorf("p2p: blocktxn: short payload")
	}
	var h [32]byte
	copy(h[:], b[:32])
	off := 32
	n, used, err := readCompactSize(b[off:])
	if err != nil {
		return nil, err
	}
	off += used
	txs := make([][]byte, 0, int(n))
	for i := 0; i < int(n); i++ {
		_, usedBytes, err := consensus.ParseTxBytesPrefix(b[off:])
		if err != nil {
			return nil, err
		}
		txb := append([]byte(nil), b[off:off+usedBytes]...)
		off += usedBytes
		txs = append(txs, txb)
	}
	if off != len(b) {
		return nil, fmt.Errorf("p2p: blocktxn: trailing bytes")
	}
	return &BlockTxnPayload{BlockHash: h, Txs: txs}, nil
}

func cmpctKeys(p crypto.CryptoProvider, header consensus.BlockHeader, nonce uint64) (uint64, uint64, error) {
	if p == nil {
		return 0, 0, fmt.Errorf("p2p: cmpct: nil crypto provider")
	}
	hb := consensus.BlockHeaderBytes(header)
	var nonce8 [8]byte
	binary.LittleEndian.PutUint64(nonce8[:], nonce)
	buf := make([]byte, 0, len(compactBlockKeyDomain)+len(hb)+len(nonce8))
	buf = append(buf, []byte(compactBlockKeyDomain)...)
	buf = append(buf, hb...)
	buf = append(buf, nonce8[:]...)
	km, err := p.SHA3_256(buf)
	if err != nil {
		return 0, 0, err
	}
	k0 := binary.LittleEndian.Uint64(km[0:8])
	k1 := binary.LittleEndian.Uint64(km[8:16])
	return k0, k1, nil
}

func ShortID(p crypto.CryptoProvider, header consensus.BlockHeader, nonce uint64, txBytes []byte) ([CompactBlockShortIDBytes]byte, error) {
	k0, k1, err := cmpctKeys(p, header, nonce)
	if err != nil {
		return [CompactBlockShortIDBytes]byte{}, err
	}
	wtxid, err := p.SHA3_256(txBytes)
	if err != nil {
		return [CompactBlockShortIDBytes]byte{}, err
	}
	s64 := siphash24(k0, k1, wtxid[:])
	var tmp8 [8]byte
	binary.LittleEndian.PutUint64(tmp8[:], s64)
	var out [CompactBlockShortIDBytes]byte
	copy(out[:], tmp8[:CompactBlockShortIDBytes])
	return out, nil
}

func rotl64(x uint64, b uint) uint64 {
	return (x << b) | (x >> (64 - b))
}

func sipround(v0, v1, v2, v3 uint64) (uint64, uint64, uint64, uint64) {
	v0 += v1
	v1 = rotl64(v1, 13)
	v1 ^= v0
	v0 = rotl64(v0, 32)

	v2 += v3
	v3 = rotl64(v3, 16)
	v3 ^= v2

	v0 += v3
	v3 = rotl64(v3, 21)
	v3 ^= v0

	v2 += v1
	v1 = rotl64(v1, 17)
	v1 ^= v2
	v2 = rotl64(v2, 32)
	return v0, v1, v2, v3
}

// siphash24 implements SipHash-2-4 for shortid derivation. This is a local
// implementation to avoid extra dependencies; correctness is exercised by
// roundtrip + determinism tests and cross-client interop tests.
func siphash24(k0, k1 uint64, msg []byte) uint64 {
	v0 := uint64(0x736f6d6570736575) ^ k0
	v1 := uint64(0x646f72616e646f6d) ^ k1
	v2 := uint64(0x6c7967656e657261) ^ k0
	v3 := uint64(0x7465646279746573) ^ k1

	off := 0
	for ; off+8 <= len(msg); off += 8 {
		m := binary.LittleEndian.Uint64(msg[off : off+8])
		v3 ^= m
		v0, v1, v2, v3 = sipround(v0, v1, v2, v3)
		v0, v1, v2, v3 = sipround(v0, v1, v2, v3)
		v0 ^= m
	}

	var last uint64 = uint64(len(msg)) << 56
	rem := msg[off:]
	for i := 0; i < len(rem); i++ {
		last |= uint64(rem[i]) << (8 * i)
	}
	v3 ^= last
	v0, v1, v2, v3 = sipround(v0, v1, v2, v3)
	v0, v1, v2, v3 = sipround(v0, v1, v2, v3)
	v0 ^= last

	v2 ^= 0xff
	for i := 0; i < 4; i++ {
		v0, v1, v2, v3 = sipround(v0, v1, v2, v3)
	}
	return v0 ^ v1 ^ v2 ^ v3
}

File: clients/rust/crates/rubin-p2p/src/compactblock.rs
use crate::compactsize;
use rubin_consensus::{
    block_header_bytes, parse_block_header_bytes, parse_tx_bytes, parse_tx_bytes_prefix,
    BlockHeader,
};
use rubin_crypto::CryptoProvider;
use siphasher::sip::SipHasher24;
use std::hash::Hasher;

pub const CMD_SENDCMPCT: &str = "sendcmpct";
pub const CMD_CMPCTBLOCK: &str = "cmpctblock";
pub const CMD_GETBLOCKTXN: &str = "getblocktxn";
pub const CMD_BLOCKTXN: &str = "blocktxn";

pub const SHORTID_BYTES: usize = 6;
const KEY_DOMAIN: &[u8] = b"RUBIN-CMPCT-v1";

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SendCmpctPayload {
    pub announce: u8,
    pub shortid_wtxid: u8, // MUST be 1 in RUBIN
    pub protocol_version: u32,
}

pub fn encode_sendcmpct(p: &SendCmpctPayload) -> Result<Vec<u8>, String> {
    if p.shortid_wtxid != 1 {
        return Err("p2p: sendcmpct: shortid_wtxid must be 1".into());
    }
    let mut out = Vec::with_capacity(6);
    out.push(p.announce);
    out.push(p.shortid_wtxid);
    out.extend_from_slice(&p.protocol_version.to_le_bytes());
    Ok(out)
}

pub fn decode_sendcmpct(b: &[u8]) -> Result<SendCmpctPayload, String> {
    if b.len() != 6 {
        return Err("p2p: sendcmpct: length mismatch".into());
    }
    let announce = b[0];
    let shortid_wtxid = b[1];
    if shortid_wtxid != 1 {
        return Err("p2p: sendcmpct: shortid_wtxid must be 1".into());
    }
    let mut tmp4 = [0u8; 4];
    tmp4.copy_from_slice(&b[2..6]);
    let protocol_version = u32::from_le_bytes(tmp4);
    Ok(SendCmpctPayload {
        announce,
        shortid_wtxid,
        protocol_version,
    })
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PrefilledTx {
    pub index: u64,
    pub tx_bytes: Vec<u8>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CmpctBlockPayload {
    pub header: BlockHeader,
    pub nonce: u64,
    pub tx_count: u64,
    pub shortids: Vec<[u8; SHORTID_BYTES]>,
    pub prefilled: Vec<PrefilledTx>,
}

pub fn encode_cmpctblock(p: &CmpctBlockPayload) -> Result<Vec<u8>, String> {
    if p.tx_count == 0 {
        return Err("p2p: cmpctblock: tx_count must be >= 1".into());
    }
    if (p.shortids.len() as u64) + (p.prefilled.len() as u64) != p.tx_count {
        return Err("p2p: cmpctblock: shortid_count+prefilled_count must equal tx_count".into());
    }

    let hb = block_header_bytes(&p.header);
    let mut out = Vec::new();
    out.extend_from_slice(&hb);
    out.extend_from_slice(&p.nonce.to_le_bytes());
    out.extend_from_slice(&compactsize::encode(p.tx_count));
    out.extend_from_slice(&compactsize::encode(p.shortids.len() as u64));
    for sid in &p.shortids {
        out.extend_from_slice(sid);
    }
    out.extend_from_slice(&compactsize::encode(p.prefilled.len() as u64));

    let mut prev_idx: u64 = 0;
    for (i, pf) in p.prefilled.iter().enumerate() {
        if pf.index >= p.tx_count {
            return Err("p2p: cmpctblock: prefilled index out of range".into());
        }
        if i == 0 {
            out.extend_from_slice(&compactsize::encode(pf.index));
        } else {
            if pf.index <= prev_idx {
                return Err("p2p: cmpctblock: prefilled indices not strictly increasing".into());
            }
            out.extend_from_slice(&compactsize::encode(pf.index - prev_idx - 1));
        }
        prev_idx = pf.index;
        // Ensure bytes are parseable to avoid ambiguity.
        let _ = parse_tx_bytes(&pf.tx_bytes)?;
        out.extend_from_slice(&pf.tx_bytes);
    }
    Ok(out)
}

pub fn decode_cmpctblock(b: &[u8]) -> Result<CmpctBlockPayload, String> {
    if b.len() < 116 + 8 + 1 {
        return Err("p2p: cmpctblock: short payload".into());
    }
    let mut off = 0usize;
    let header = parse_block_header_bytes(&b[0..116])?;
    off += 116;
    let mut tmp8 = [0u8; 8];
    tmp8.copy_from_slice(&b[off..off + 8]);
    let nonce = u64::from_le_bytes(tmp8);
    off += 8;

    let (tx_count, used) = compactsize::decode(&b[off..])?;
    off += used;
    if tx_count < 1 {
        return Err("p2p: cmpctblock: tx_count must be >= 1".into());
    }

    let (shortid_count, used) = compactsize::decode(&b[off..])?;
    off += used;
    if shortid_count > tx_count {
        return Err("p2p: cmpctblock: shortid_count exceeds tx_count".into());
    }
    let need_sids = (shortid_count as usize)
        .checked_mul(SHORTID_BYTES)
        .ok_or_else(|| "p2p: cmpctblock: shortid_count overflow".to_string())?;
    if b.len() < off + need_sids {
        return Err("p2p: cmpctblock: shortids truncated".into());
    }
    let mut shortids = Vec::with_capacity(shortid_count as usize);
    for _ in 0..(shortid_count as usize) {
        let mut sid = [0u8; SHORTID_BYTES];
        sid.copy_from_slice(&b[off..off + SHORTID_BYTES]);
        off += SHORTID_BYTES;
        shortids.push(sid);
    }

    let (prefilled_count, used) = compactsize::decode(&b[off..])?;
    off += used;
    if prefilled_count > tx_count {
        return Err("p2p: cmpctblock: prefilled_count exceeds tx_count".into());
    }
    if prefilled_count + shortid_count != tx_count {
        return Err("p2p: cmpctblock: shortid_count+prefilled_count must equal tx_count".into());
    }

    let mut prefilled = Vec::with_capacity(prefilled_count as usize);
    let mut prev_idx: u64 = 0;
    for i in 0..(prefilled_count as usize) {
        let (delta, u) = compactsize::decode(&b[off..])?;
        off += u;
        let idx = if i == 0 { delta } else { prev_idx + 1 + delta };
        if idx >= tx_count {
            return Err("p2p: cmpctblock: prefilled index out of range".into());
        }
        if i > 0 && idx <= prev_idx {
            return Err("p2p: cmpctblock: prefilled indices not strictly increasing".into());
        }
        prev_idx = idx;

        let (_tx, used_bytes) = parse_tx_bytes_prefix(&b[off..])?;
        let tx_bytes = b[off..off + used_bytes].to_vec();
        off += used_bytes;
        prefilled.push(PrefilledTx {
            index: idx,
            tx_bytes,
        });
    }

    if off != b.len() {
        return Err("p2p: cmpctblock: trailing bytes".into());
    }

    Ok(CmpctBlockPayload {
        header,
        nonce,
        tx_count,
        shortids,
        prefilled,
    })
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct GetBlockTxnPayload {
    pub block_hash: [u8; 32],
    pub indices: Vec<u64>,
}

pub fn encode_getblocktxn(p: &GetBlockTxnPayload) -> Result<Vec<u8>, String> {
    if p.indices.is_empty() {
        return Err("p2p: getblocktxn: empty indices".into());
    }
    let mut out = Vec::new();
    out.extend_from_slice(&p.block_hash);
    out.extend_from_slice(&compactsize::encode(p.indices.len() as u64));
    let mut prev = 0u64;
    for (i, idx) in p.indices.iter().enumerate() {
        if i == 0 {
            out.extend_from_slice(&compactsize::encode(*idx));
        } else {
            if *idx <= prev {
                return Err("p2p: getblocktxn: indices not strictly increasing".into());
            }
            out.extend_from_slice(&compactsize::encode(*idx - prev - 1));
        }
        prev = *idx;
    }
    Ok(out)
}

pub fn decode_getblocktxn(b: &[u8]) -> Result<GetBlockTxnPayload, String> {
    if b.len() < 32 + 1 {
        return Err("p2p: getblocktxn: short payload".into());
    }
    let mut block_hash = [0u8; 32];
    block_hash.copy_from_slice(&b[0..32]);
    let mut off = 32usize;
    let (n, used) = compactsize::decode(&b[off..])?;
    off += used;
    if n == 0 {
        return Err("p2p: getblocktxn: empty indices".into());
    }
    let mut indices = Vec::with_capacity(n as usize);
    let mut prev = 0u64;
    for i in 0..(n as usize) {
        let (delta, u) = compactsize::decode(&b[off..])?;
        off += u;
        let idx = if i == 0 { delta } else { prev + 1 + delta };
        if i > 0 && idx <= prev {
            return Err("p2p: getblocktxn: indices not strictly increasing".into());
        }
        prev = idx;
        indices.push(idx);
    }
    if off != b.len() {
        return Err("p2p: getblocktxn: trailing bytes".into());
    }
    Ok(GetBlockTxnPayload {
        block_hash,
        indices,
    })
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BlockTxnPayload {
    pub block_hash: [u8; 32],
    pub txs: Vec<Vec<u8>>,
}

pub fn encode_blocktxn(p: &BlockTxnPayload) -> Result<Vec<u8>, String> {
    let mut out = Vec::new();
    out.extend_from_slice(&p.block_hash);
    out.extend_from_slice(&compactsize::encode(p.txs.len() as u64));
    for txb in &p.txs {
        let _ = parse_tx_bytes(txb)?;
        out.extend_from_slice(txb);
    }
    Ok(out)
}

pub fn decode_blocktxn(b: &[u8]) -> Result<BlockTxnPayload, String> {
    if b.len() < 32 + 1 {
        return Err("p2p: blocktxn: short payload".into());
    }
    let mut block_hash = [0u8; 32];
    block_hash.copy_from_slice(&b[0..32]);
    let mut off = 32usize;
    let (n, used) = compactsize::decode(&b[off..])?;
    off += used;
    let mut txs = Vec::with_capacity(n as usize);
    for _ in 0..(n as usize) {
        let (_tx, used_bytes) = parse_tx_bytes_prefix(&b[off..])?;
        txs.push(b[off..off + used_bytes].to_vec());
        off += used_bytes;
    }
    if off != b.len() {
        return Err("p2p: blocktxn: trailing bytes".into());
    }
    Ok(BlockTxnPayload { block_hash, txs })
}

fn cmpct_keys(
    provider: &dyn CryptoProvider,
    header: &BlockHeader,
    nonce: u64,
) -> Result<(u64, u64), String> {
    let hb = block_header_bytes(header);
    let mut buf = Vec::with_capacity(KEY_DOMAIN.len() + hb.len() + 8);
    buf.extend_from_slice(KEY_DOMAIN);
    buf.extend_from_slice(&hb);
    buf.extend_from_slice(&nonce.to_le_bytes());
    let km = provider.sha3_256(&buf)?;
    let mut a = [0u8; 8];
    let mut b8 = [0u8; 8];
    a.copy_from_slice(&km[0..8]);
    b8.copy_from_slice(&km[8..16]);
    Ok((u64::from_le_bytes(a), u64::from_le_bytes(b8)))
}

pub fn shortid(
    provider: &dyn CryptoProvider,
    header: &BlockHeader,
    nonce: u64,
    tx_bytes: &[u8],
) -> Result<[u8; SHORTID_BYTES], String> {
    let (k0, k1) = cmpct_keys(provider, header, nonce)?;
    let wtxid = provider.sha3_256(tx_bytes)?;
    let mut hasher = SipHasher24::new_with_keys(k0, k1);
    hasher.write(&wtxid);
    let s64 = hasher.finish();
    let le = s64.to_le_bytes();
    let mut out = [0u8; SHORTID_BYTES];
    out.copy_from_slice(&le[0..SHORTID_BYTES]);
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use rubin_consensus::{
        tx_bytes, Tx, TxInput, TxOutput, WitnessSection, TX_KIND_STANDARD, TX_VERSION_V2,
    };
    use rubin_crypto::DevStdCryptoProvider;

    #[test]
    fn sendcmpct_roundtrip() {
        let p = SendCmpctPayload {
            announce: 1,
            shortid_wtxid: 1,
            protocol_version: 1,
        };
        let raw = encode_sendcmpct(&p).unwrap();
        let dec = decode_sendcmpct(&raw).unwrap();
        assert_eq!(p, dec);
    }

    #[test]
    fn cmpctblock_roundtrip() {
        let header = BlockHeader {
            version: 1,
            prev_block_hash: [0u8; 32],
            merkle_root: [0u8; 32],
            timestamp: 1,
            target: [0u8; 32],
            nonce: 2,
        };
        let tx = Tx {
            version: TX_VERSION_V2,
            tx_kind: TX_KIND_STANDARD,
            tx_nonce: 1,
            inputs: vec![TxInput {
                prev_txid: [0u8; 32],
                prev_vout: 0,
                script_sig: vec![],
                sequence: 0,
            }],
            outputs: vec![TxOutput {
                value: 0,
                covenant_type: 0,
                covenant_data: vec![],
            }],
            locktime: 0,
            da_commit: None,
            da_chunk: None,
            da_payload: vec![],
            witness: WitnessSection { witnesses: vec![] },
        };
        let txb = tx_bytes(&tx);
        let payload = CmpctBlockPayload {
            header,
            nonce: 42,
            tx_count: 2,
            shortids: vec![[1, 2, 3, 4, 5, 6]],
            prefilled: vec![PrefilledTx {
                index: 0,
                tx_bytes: txb.clone(),
            }],
        };
        let raw = encode_cmpctblock(&payload).unwrap();
        let dec = decode_cmpctblock(&raw).unwrap();
        assert_eq!(dec.tx_count, 2);
        assert_eq!(dec.shortids.len(), 1);
        assert_eq!(dec.prefilled.len(), 1);
        assert_eq!(dec.prefilled[0].tx_bytes, txb);
    }

    #[test]
    fn shortid_is_deterministic() {
        let provider = DevStdCryptoProvider;
        let header = BlockHeader {
            version: 1,
            prev_block_hash: [0u8; 32],
            merkle_root: [0u8; 32],
            timestamp: 1,
            target: [0u8; 32],
            nonce: 2,
        };
        let tx = Tx {
            version: TX_VERSION_V2,
            tx_kind: TX_KIND_STANDARD,
            tx_nonce: 1,
            inputs: vec![TxInput {
                prev_txid: [0u8; 32],
                prev_vout: 0,
                script_sig: vec![],
                sequence: 0,
            }],
            outputs: vec![TxOutput {
                value: 0,
                covenant_type: 0,
                covenant_data: vec![],
            }],
            locktime: 0,
            da_commit: None,
            da_chunk: None,
            da_payload: vec![],
            witness: WitnessSection { witnesses: vec![] },
        };
        let txb = tx_bytes(&tx);
        let a = shortid(&provider, &header, 123, &txb).unwrap();
        let b = shortid(&provider, &header, 123, &txb).unwrap();
        assert_eq!(a, b);
    }
}
